src/kernel-install/ukify.py

```diff
@@ -0,0 +1,187 @@
+            p = pathlib.Path('/usr/lib/os-release')
+        args.os_release = p
+
+    if args.efi_arch is None:
+        args.efi_arch = guess_efi_arch()
+
+    if args.stub is None:
+        args.stub = f'/usr/lib/systemd/boot/efi/linux{args.efi_arch}.efi.stub'
+
+    if args.output is None:
+        suffix = '.efi' if args.key else '.unsigned.efi'
+        args.output = args.linux.name + suffix
+
+    return args
+
+
+def check_inputs(args):
+    for name, value in vars(args).items():
+        if name == 'output':
+            continue
+
+        if not isinstance(value, pathlib.Path):
+            continue
+
+        # Open file to check that we can read it, or generate an exception
+        open(value).close()
+
+
+def make_uki(args):
+    files = []  # a holder for temporary files
+    sections = []
+
+    def add_section(name, content, offset):
+        if not content:
+            return
+
+        if isinstance(content, str):
+            tmp = tempfile.NamedTemporaryFile(mode='wt', prefix=f'tmp{name}')
+            tmp.write(content)
+            tmp.flush()
+            files.append(tmp)
+            content = tmp.name
+
+        sections.append((name, content, offset))
+
+    add_section('osrel',   args.os_release,    0x20_000)
+    add_section('cmdline', args.cmdline,       0x30_000)
+    add_section('dtb',     args.devicetree,    0x40_000)
+    add_section('splash',  args.splash,       0x100_000)
+    add_section('linux',   args.linux,      0x2_000_000)
+    add_section('initrd',  args.initrd,     0x3_000_000)
+
+    if args.key:
+        unsigned = tempfile.NamedTemporaryFile(prefix='linux')
+        output = unsigned.name
+    else:
+        output = args.output
+
+    objcopy = [
+        'objcopy',
+        args.stub,
+        *itertools.chain.from_iterable(
+            ('--add-section',        f'.{name}={content}',
+             '--change-section-vma', f'.{name}=0x{offset:x}')
+            for name, content, offset in sections),
+        output,
+    ]
+    print('+', shell_join(objcopy))
+    subprocess.check_call(objcopy)
+
+    # signing
+
+    if args.key:
+        sbsign = [
+            'sbsign',
+            '--key', args.key,
+            '--cert', args.cert,
+            unsigned.name,
+            '--output', args.output,
+        ]
+        print('+', shell_join(sbsign))
+        subprocess.check_call(sbsign)
+
+        # We end up with no executable bits, let's reapply them
+        os.umask(umask := os.umask(0))
+        os.chmod(args.output, 0o777 & ~umask)
+
+    print(f"Wrote {'signed' if args.key else 'unsigned'} {args.output}")
+
+    # measurement
+
+    if args.measure:
+        measure = [
+            '/usr/lib/systemd/systemd-measure',
+            'calculate',
+            *itertools.chain.from_iterable(
+                (f'--{name}', content)
+                for name, content, offset in sections),
+        ]
+        print('+', shell_join(measure))
+        subprocess.check_call(measure)
```

Comment:
hmm, the idea is to calculate the expected PCR values and *sign* them, before you link things together with objcopy.

i.e.

`systemd-measure sign` → `objcopy` → `sbsign`

Reply:
Indeed. The code does that now.